File: Javascript/Structural Patterns/Facade/facade.ts
class AudioSystem {
    volume: number
    speed: number
    
    constructor() {
        this.volume = 50
        this.speed = 1
    }

    setSpeed(level: number) {
        this.speed = level
        console.log(`[Audio] >>> Speed set to ${level}`)
    }

    setVolume(level: number) {
        this.volume = level
        console.log(`[Audio] >>> Volume set to ${level}`)
    }
}

class VideoSystem {
    speed: number
    width: number
    height: number

    constructor() {
        this.speed = 1
        this.width = 1024
        this.height = 1024
    }

    setSpeed(level: number) {
        this.speed = level
        console.log(`[Video] >>> Speed set to ${level}`)
    }

    setResolution(width: number, height: number) {
        this.width = width
        this.height = height
        console.log(`[Video] >>> Resolution set to ${width}x${height}`)
    }
}

class NotificationSystem {
    receivedNotification: boolean

    constructor() {
        this.receivedNotification = true
    }

    setNotification(state: boolean) {
        this.receivedNotification = state
        console.log(`[Notification] >>> Notification state set to ${state}`)
    }
}

export class MediaFacade {
    private audio: AudioSystem
    private video: VideoSystem
    private notification: NotificationSystem

    constructor() {
        this.audio = new AudioSystem()
        this.video = new VideoSystem()
        this.notification = new NotificationSystem()
    }

    play() {
        console.log("[Facade] >>> Play video")
        this.audio.setSpeed(1)
        this.audio.setVolume(60)
        this.video.setSpeed(1)
        this.video.setResolution(1440, 1440)
    }

    turnOffNotification() {
        console.log("[Facade] >>> Turn off notification")
        this.notification.setNotification(false)
    }
}

File: Go/Structural_Patterns/Facade/facade.go
package facade

import "fmt"

type MediaFacade struct {
	Audio        *AudioSystem
	Video        *VideoSystem
	Notification *NotificationSystem
}

func (m *MediaFacade) Play() {
	fmt.Println("[Facade] >>> Play video")
	m.Audio.setSpeed(1)
	m.Audio.setVolume(60)
	m.Video.setSpeed(1)
	m.Video.setResolution(1440, 1440)
}

func NewMediaFacade() *MediaFacade {
	return &MediaFacade{
		Audio:        newAudioSystem(),
		Video:        newVideoSystem(),
		Notification: newNotificationSystem(),
	}
}

type AudioSystem struct {
	volume, speed int
}

func (a *AudioSystem) setSpeed(speed int) {
	fmt.Printf("Audio Speed set to %d\n", speed)
	a.speed = speed
}

func (a *AudioSystem) setVolume(volume int) {
	fmt.Printf("Audio Volume set to %d\n", volume)
	a.volume = volume
}

func newAudioSystem() *AudioSystem {
	return &AudioSystem{
		volume: 50,
		speed:  1,
	}
}

type VideoSystem struct {
	speed, width, height int
}

func (v *VideoSystem) setSpeed(speed int) {
	fmt.Printf("Video Speed set to %d\n", speed)
	v.speed = speed
}

func (v *VideoSystem) setResolution(width int, height int) {
	fmt.Printf("Video Resolution set to %d, %d\n", width, height)
	v.width = width
	v.height = height
}

func newVideoSystem() *VideoSystem {
	return &VideoSystem{
		speed:  1,
		width:  1024,
		height: 1024,
	}
}

type NotificationSystem struct {
	receivedNotification bool
}

func (n *NotificationSystem) setNotification(notification bool) {
	fmt.Printf("Notification set to %d\n", notification)
	n.receivedNotification = notification
}

func newNotificationSystem() *NotificationSystem {
	return &NotificationSystem{
		receivedNotification: true,
	}
}
